relbench/tasks/f1.py

```diff
@@ -0,0 +1,311 @@
+import duckdb
+import numpy as np
+import pandas as pd
+from tqdm import tqdm
+
+from relbench.data import Database, RelBenchTask, Table
+from relbench.data.task import TaskType
+from relbench.metrics import accuracy, average_precision, f1, mae, rmse, roc_auc
+from relbench.utils import get_df_in_window
+
+
+class PointsTask(RelBenchTask):
+    r"""Predict the finishing position of each driver in a race."""
+    name = "rel-f1-points"
+    task_type = TaskType.REGRESSION  # TaskType.BINARY_CLASSIFICATION
+    entity_col = "driverId"
+    entity_table = "drivers"
+    time_col = "date"
+    target_col = "points"
+    timedelta = pd.Timedelta(days=30)
+    metrics = [mae, rmse]  # [average_precision, accuracy, f1, roc_auc]
+
+    def make_table(self, db: Database, timestamps: "pd.Series[pd.Timestamp]") -> Table:
+        r"""Create Task object for results_position_next_race."""
+        timestamp_df = pd.DataFrame({"timestamp": timestamps})
+
+        results = db.table_dict["results"].df
+        drivers = db.table_dict["drivers"].df
+        races = db.table_dict["races"].df
+
+        df = duckdb.sql(
+            f"""
+                SELECT
+                    t.timestamp as date,
+                    dri.driverId as driverId,
+                    sum(re.points) points
+                FROM
+                    timestamp_df t
+                LEFT JOIN
+                    results re
+                ON
+                    re.date <= t.timestamp + INTERVAL '{self.timedelta}'
+                    and re.date  > t.timestamp
+                LEFT JOIN
+                    drivers dri
+                ON
+                    re.driverId = dri.driverId
+                WHERE
+                    dri.driverId IN (
+                        SELECT DISTINCT driverId
+                        FROM results
+                        WHERE date > t.timestamp - INTERVAL '1 year'
+                    )
+                GROUP BY t.timestamp, dri.driverId
+
+            ;
+            """
+        ).df()
+
+        # make into binary classification task
+        # df[self.target_col] = df[self.target_col].apply(lambda x: 1 if x > 0. else 0)
+
+        mean = 3.021777777777778
+        std = 5.299464335614526
+        df[self.target_col] = (df[self.target_col] - mean) / std
+
+        return Table(
+            df=df,
+            fkey_col_to_pkey_table={self.entity_col: self.entity_table},
+            pkey_col=None,
+            time_col=self.time_col,
+        )
```

Comment:
Usually, `make_table` is enough if you inherit `RelBenchTask`. Could you clarify why you also add `train_table`, `val_table`, `test_table`?

Reply:
train_table is usually collected using data from only one timestamp. In the case of F1 drivers, one timestamp means you get a test set of size only ~20 (since this is about how many drivers are competing at any given time). So in order to build a test set that is big enough to be reliable I had to compute test data over many timestamps. Similarly for validation data.